File: py-sdk/src/mcp/types/annotations.py
"""``Annotations`` — optional, untrusted hints on content blocks and resources (§14.6).

Trust model (R-14.6-f/-g): consumers MUST NOT use annotation values for security or
correctness decisions; they are advisory only and MAY influence presentation, ordering,
or context-inclusion.
"""

from __future__ import annotations

from mcp.types.role import is_role


def _is_number(value: object) -> bool:
  return isinstance(value, (int, float)) and not isinstance(value, bool)


def is_valid_annotations(value: object) -> bool:
  """Return ``True`` for a valid ``Annotations`` object (§14.6).

  All fields OPTIONAL (an empty object is valid); extra members are tolerated:

  * ``audience`` — list of ``Role`` values (R-14.6-b);
  * ``priority`` — number in the inclusive range ``0..1`` (R-14.6-c, R-14.6-d);
  * ``lastModified`` — ISO-8601 timestamp string (R-14.6-e).
  """
  if not isinstance(value, dict):
    return False
  if "audience" in value:
    audience = value["audience"]
    if not isinstance(audience, list) or not all(is_role(r) for r in audience):
      return False
  if "priority" in value:
    priority = value["priority"]
    if not _is_number(priority) or not (0 <= priority <= 1):
      return False
  if "lastModified" in value and not isinstance(value["lastModified"], str):
    return False
  return True

File: csharp-sdk/Protocol/Annotations.cs
using System.Text.Json.Serialization;

namespace Stackific.Mcp.Protocol;

/// <summary>
/// The author/recipient role of a message or the intended audience of content (spec §14.7).
/// The only permitted wire values are <c>user</c> and <c>assistant</c>.
/// </summary>
[JsonConverter(typeof(JsonStringEnumConverter<Role>))]
public enum Role
{
  /// <summary>The human participant.</summary>
  [JsonStringEnumMemberName("user")]
  User,

  /// <summary>The language-model participant.</summary>
  [JsonStringEnumMemberName("assistant")]
  Assistant,
}

/// <summary>
/// Optional, untrusted hints about a piece of content or a resource (spec §14.6). A consumer
/// MUST NOT rely on annotation values for security or correctness decisions.
/// </summary>
public sealed record Annotations
{
  /// <summary>OPTIONAL. The intended audience(s) for the annotated object.</summary>
  public IReadOnlyList<Role>? Audience { get; init; }

  /// <summary>OPTIONAL. Importance from 0 (least) to 1 (most), inclusive.</summary>
  public double? Priority { get; init; }

  /// <summary>OPTIONAL. ISO-8601 timestamp of the last modification.</summary>
  public string? LastModified { get; init; }
}
